File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0;
        std::unordered_set<int> dupset;
        int maxlen = 0;

        for (int r = 0; r < s.size(); r++) {
            while (dupset.contains(s[r])) {
                dupset.erase(s[l]);
                l += 1;
            }
            dupset.insert(s[r]);
            maxlen = std::max(maxlen, r - l + 1);
        }

        return maxlen;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count_dict = {}
        l, r = 0, 0
        maxfreq = 0
        maxlen = 0

        while r < len(s): 
            count_dict[s[r]] = 1 + count_dict.get(s[r], 0)
            maxfreq = max(maxfreq, count_dict[s[r]])

            while (r - l + 1) - maxfreq > k:
                count_dict[s[l]] -= 1
                l += 1

            maxlen = max(maxlen, r - l + 1)

            r += 1

        return maxlen